nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Using speed limit in m/s
+    percentage_ = false;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Speed filter is using absolute speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else if (msg->type == SPEED_FILTER_PERCENT) {
+    // Using speed limit in % of maximum speed
+    percentage_ = true;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Speed filter is using expressed in a percent from maximum speed"
+      "speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else {
+    RCLCPP_ERROR(node->get_logger(), "Mode is not supported by SpeedFilter");
+    throw std::runtime_error("Mode is not supported by SpeedFilter");
+  }
+
+  // Reset speed limit when got new filter info
+  speed_limit_ = NO_SPEED_LIMIT;
+  speed_limit_prev_ = FAKE_SPEED_LIMIT;  // Fake speed limit to force speed limit topic update
+
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void SpeedFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+bool SpeedFilter::transformPose(
+  const geometry_msgs::msg::Pose2D & pose,
+  geometry_msgs::msg::Pose2D & mask_pose) const
+{
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
```

Comment:
Please just store logger in `logger_` and use that so you don't need to get the node and lock it every time just to get a logger. That's the new design paradigm we have for all the plugins after the last set of PRs to fix shutdown.

Reply:
Done in https://github.com/ros-planning/navigation2/pull/2074/commits/3465c90415afd11418f98fc01bbc9386afa853ea